File: app/src/main/java/com/tichuguru/StatsListFragment.java
package com.tichuguru;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.tichuguru.model.Game;
import com.tichuguru.model.Player;
import java.util.List;

public class StatsListFragment extends Fragment {
    private Player player;

    public static StatsListFragment newInstance(String title, String[] labels, String[] values, String playerName) {
        StatsListFragment f = new StatsListFragment();
        Bundle args = new Bundle();
        args.putString("title", title);
        args.putSerializable("labels", labels);
        args.putSerializable("values", values);
        if (playerName != null) args.putString("playerName", playerName);
        f.setArguments(args);
        return f;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        String playerName = requireArguments().getString("playerName");
        player = playerName != null ? TGApp.getPlayer(playerName) : null;
        return inflater.inflate(player == null ? R.layout.rankinglist : R.layout.statslist, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Bundle args = requireArguments();
        requireActivity().setTitle(args.getString("title"));

        String[] labels = (String[]) args.getSerializable("labels");
        String[] values = (String[]) args.getSerializable("values");

        RecyclerView statsList = view.findViewById(R.id.statsList);
        statsList.setLayoutManager(new LinearLayoutManager(requireContext()));
        statsList.setAdapter(new StatsAdapter(labels, values));

        if (player != null) {
            view.findViewById(R.id.statsClearPlayerStats).setOnClickListener(v -> onClearStats());
            view.findViewById(R.id.statsDelPlayer).setOnClickListener(v -> onDeletePlayer());
        }
    }

    private void onClearStats() {
        new AlertDialog.Builder(requireContext())
            .setMessage("Are you sure?")
            .setPositiveButton("Yes", (dialog, which) -> {
                player.clearStats();
                ((TGApp) requireActivity().getApplication()).savePlayers();
                new ViewModelProvider(requireActivity()).get(TGViewModel.class).notifyPlayersChanged();
                getParentFragmentManager().popBackStack();
            })
            .setNegativeButton("No", null)
            .show();
    }

    private void onDeletePlayer() {
        new AlertDialog.Builder(requireContext())
            .setMessage("Are you sure?")
            .setPositiveButton("Yes", (dialog, which) -> {
                boolean updateSharedGame = false;
                List<Game> games = TGApp.getGames();
                for (int i = games.size() - 1; i >= 0; i--) {
                    Game game = games.get(i);
                    if (game.containsPlayer(player)) {
                        games.remove(i);
                        if (game == TGApp.getGame()) updateSharedGame = true;
                    }
                }
                if (updateSharedGame) {
                    TGApp.setGame(games.isEmpty() ? null : games.get(games.size() - 1));
                }
                TGApp.getPlayers().remove(player);
                TGApp app = (TGApp) requireActivity().getApplication();
                app.saveGames();
                app.savePlayers();
                TGViewModel vm = new ViewModelProvider(requireActivity()).get(TGViewModel.class);
                vm.notifyGamesChanged();
                vm.notifyPlayersChanged();
                getParentFragmentManager().popBackStack();
            })
            .setNegativeButton("No", null)
            .show();
    }

    private static class StatsAdapter extends RecyclerView.Adapter<StatsAdapter.ViewHolder> {
        private final String[] labels;
        private final String[] values;

        StatsAdapter(String[] labels, String[] values) {
            this.labels = labels;
            this.values = values;
        }

        static class ViewHolder extends RecyclerView.ViewHolder {
            TextView label, value;

            ViewHolder(View v) {
                super(v);
                label = v.findViewById(R.id.statsLabel);
                value = v.findViewById(R.id.statsValue);
            }
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.statslistrow, parent, false);
            return new ViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            holder.label.setText(labels[position]);
            if (values[position] != null) {
                holder.label.setTypeface(null, 0);
                holder.label.setTextSize(18.0f);
                holder.value.setText(values[position]);
            } else {
                holder.label.setTypeface(null, 1);
                holder.label.setTextSize(24.0f);
                holder.value.setText("");
            }
        }

        @Override
        public int getItemCount() { return labels.length; }
    }
}

File: app/src/main/kotlin/com/tichuguru/StatsListFragment.kt
package com.tichuguru

import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tichuguru.databinding.StatslistrowBinding
import com.tichuguru.model.Player

class StatsListFragment : Fragment() {
    private var player: Player? = null
    private lateinit var viewModel: TGViewModel

    companion object {
        private const val ARG_PLAYER_ID = "playerId"

        fun newInstance(
            title: String,
            labels: Array<String>,
            values: Array<String?>,
            player: Player?,
        ): StatsListFragment =
            StatsListFragment().apply {
                arguments =
                    Bundle().apply {
                        putString("title", title)
                        putStringArray("labels", labels)
                        @Suppress("UNCHECKED_CAST")
                        putStringArray("values", values as Array<String>)
                        if (player != null) putLong(ARG_PLAYER_ID, player.dbId)
                    }
            }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        viewModel = ViewModelProvider(requireActivity())[TGViewModel::class.java]
        val playerId = requireArguments().getLong(ARG_PLAYER_ID, -1L)
        player = if (playerId != -1L) viewModel.getPlayerById(playerId) else null
        return inflater.inflate(if (player == null) R.layout.rankinglist else R.layout.statslist, container, false)
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        val args = requireArguments()
        requireActivity().title = args.getString("title")

        val labels = requireNotNull(args.getStringArray("labels")) { "labels arg missing" }

        @Suppress("UNCHECKED_CAST")
        val values = args.getStringArray("values") as Array<String?>

        val statsList = view.findViewById<RecyclerView>(R.id.statsList)
        statsList.layoutManager = LinearLayoutManager(requireContext())
        statsList.adapter = StatsAdapter(labels, values)

        if (player != null) {
            view.findViewById<View>(R.id.statsRenamePlayer).setOnClickListener { onRenamePlayer() }
            view.findViewById<View>(R.id.statsClearPlayerStats).setOnClickListener { onClearStats() }
            view.findViewById<View>(R.id.statsDelPlayer).setOnClickListener { onDeletePlayer() }
        }
    }

    private fun onRenamePlayer() {
        val p = checkNotNull(player) { "player not set" }
        val ctx = requireContext()
        val activity = requireActivity()
        val input = EditText(ctx)
        input.setText(p.name)
        input.selectAll()
        AlertDialog
            .Builder(ctx)
            .setTitle("Rename Player")
            .setView(input)
            .setPositiveButton("Ok") { _, _ ->
                if (!isAdded) return@setPositiveButton
                val newName = input.text.toString().trim()
                when {
                    newName.isEmpty() -> {
                        AlertDialog.Builder(ctx).setMessage("Name cannot be empty").show()
                    }

                    newName == p.name -> {}

                    viewModel.getPlayer(newName) != null -> {
                        AlertDialog.Builder(ctx).setMessage("A player named \"$newName\" already exists").show()
                    }

                    else -> {
                        viewModel.renamePlayer(p, newName)
                        activity.title = "Stats for $newName"
                    }
                }
            }.setNegativeButton("Cancel", null)
            .show()
    }

    private fun onClearStats() {
        AlertDialog
            .Builder(requireContext())
            .setMessage("Are you sure?")
            .setPositiveButton("Yes") { _, _ ->
                viewModel.clearPlayerStats(checkNotNull(player) { "player not set" })
                parentFragmentManager.popBackStack()
            }.setNegativeButton("No", null)
            .show()
    }

    private fun onDeletePlayer() {
        AlertDialog
            .Builder(requireContext())
            .setMessage("Are you sure?")
            .setPositiveButton("Yes") { _, _ ->
                viewModel.deletePlayer(checkNotNull(player) { "player not set" })
                parentFragmentManager.popBackStack()
            }.setNegativeButton("No", null)
            .show()
    }

    private class StatsAdapter(private val labels: Array<String>, private val values: Array<String?>) :
        RecyclerView.Adapter<StatsAdapter.ViewHolder>() {
        class ViewHolder(val binding: StatslistrowBinding) : RecyclerView.ViewHolder(binding.root)

        override fun onCreateViewHolder(
            parent: ViewGroup,
            viewType: Int,
        ): ViewHolder = ViewHolder(StatslistrowBinding.inflate(LayoutInflater.from(parent.context), parent, false))

        override fun onBindViewHolder(
            holder: ViewHolder,
            position: Int,
        ) {
            holder.binding.statsLabel.text = labels[position]
            val v = values[position]
            if (v != null) {
                holder.binding.statsLabel.setTypeface(null, Typeface.NORMAL)
                holder.binding.statsLabel.textSize = 18f
                holder.binding.statsValue.text = v
            } else {
                holder.binding.statsLabel.setTypeface(null, Typeface.BOLD)
                holder.binding.statsLabel.textSize = 24f
                holder.binding.statsValue.text = ""
            }
        }

        override fun getItemCount() = labels.size
    }
}
